rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,160 @@
+
+ALPN (Application-Layer Protocol Negotiation) is a TLS extension that allows the application layer to negotiate which protocol should be performed over a TLS connection. Teleport proxy will leverage ALPN routing in order to reduce the number of exposed ports.
+
+Outgoing traffic send to the ssh-proxy, reverse tunnel, MySQL, Postgres and MongoDB proxy services will be wrapped in TLS protocol where the teleport proxy clients (tsh, teleport internal libs, local db proxy) will be responsible for setting one of the following teleport protocols:
+- teleport-ssh-proxy
+- teleport-reverse-tunnel
+- teleport-mysql
+- teleport-postgres
+- teleport-mongodb
+
+The proxy server will listen on port 443 and will terminate the incoming TLS connection and forward the downstream traffic to the appropriate service based on the TLS ALPN negotiated protocol.
+
+A downside of this approach is that traffic will be double encrypted, once by the TLS layer, and again by SSH, MySQL and Postgres protocols. This seems like a reasonable tradeoff for the external simplicity of only needing to run a single port and protocol for all connectivity.
+
+### Reducing the number of teleport port by TLS SNI routing (Server Name Indication)
+
+In order to expose WEB, Kubernetes proxy services in one teleport proxy port, the teleport proxy server will route the incoming HTTPS traffic based on SNI TLS value either to web service or Kubernetes service.
+
+#### Setting SNI value for kubectl CLI:
+
+`tsh kube login` command generates a local kubeconfig.yaml file used during accessing teleport proxy by kubectl CLI. Additional to the current configuration the  tls-server-name  failed will be added with appropriate SNI value:
+
+```yaml
+apiVersion: v1
+clusters:
+- cluster:
+  certificate-authority-data: ...
+  server: public_proxy_ip:443
+  tls-server-name: kube.proxy.example.teleport.com
+  name: proxy.example.teleport.com
+```
+
+
+### Local teleport proxy
+
+To handle connections established to teleport proxy by external clients like mysql, psql CLIs where setting custom ALPN value is not possible we will run local teleport proxy. Traffic sent from external teleport clients will be forwarded through the local proxy where the proxy will be responsible for wrapping incoming connections in TLS protocol, setting appropriate ALPN protocol and establishing a connection to the remote teleport proxy.
+
+For the tsh ssh and tsh scp commands scope forwarding traffic through local proxy seems to be superfluous. SSH traffic can be easily wrapped in TLS with custom teleport-ssh-proxy ALPN protocol inside tsh ssh handler after detecting that the teleport proxy_listener_addr listener was enabled but we should consider adding support for ssh proxy in order to enable external SSH and SCP clients like FileZilla, WinSCP and Putty.
+
+Question:
+- Do we want to support Unix sockets in local proxy  ? 
+
+### CLI UX 
+
+#### tsh proxy connect db-instance-name
```

Comment:
Do we need a separate command though? Can we use the same `tsh db connect` I introduced in the mongo PR? Otherwise might be confusing for users to learns 2 commands and understand which one to use when.

Reply:
good idea,  I will update the RFD